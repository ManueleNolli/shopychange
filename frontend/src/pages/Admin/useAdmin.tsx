import { useToast } from '@chakra-ui/react'
import { useState, useEffect } from 'react'
import { useUserContext } from '../../context/userContext'
import {
  AdminQuery,
  salesQuery,
  handlerQueryErrorCatch,
} from '../../services/BackendService'
import {
  cleanMarketplaceStorage,
  cancelSale,
  setNewRoyalty,
  waitTransaction,
  withdrawOwner,
  withdrawTo,
  withdrawToAmount as withdrawToAmountService,
} from '../../services/BlockchainService'
import { NFT } from '../../types/components/NFT'
import useLoading from '../../hooks/useLoading'

export default function useAdmin() {
  const toast = useToast()
  const { userAddress, blockchainNetworkId } = useUserContext()
  const [royaltyValue, setRoyaltyValue] = useState<number | null>(null)
  const [marketplaceBalance, setMarketplaceBalance] = useState<number | null>(
    null
  )
  const [newRoyaltyValue, setNewRoyaltyValue] = useState<number>(0)

  const {
    isLoading: isNewRoyaltyValueLoading,
    setIsLoading: setIsNewRoyaltyValueLoading,
  } = useLoading()
  const {
    isLoading: isCleanStorageLoading,
    setIsLoading: setIsCleanStorageLoading,
  } = useLoading()
  const {
    isLoading: isCancelSaleLoading,
    setIsLoading: setIsCancelSaleLoading,
  } = useLoading()
  const { isLoading: isWithdrawLoading, setIsLoading: setIsWithdrawLoading } =
    useLoading()
  const {
    isLoading: isWithdrawToLoading,
    setIsLoading: setIsWithdrawToLoading,
  } = useLoading()
  const {
    isLoading: isWithdrawToAmountLoading,
    setIsLoading: setIsWithdrawToAmountLoading,
  } = useLoading()

  const [sales, setSales] = useState<NFT[]>([])
  const [saleToCancel, setSaleToCancel] = useState<NFT | null>(null)

  const getRoyaltyValue = async () => {
    if (!userAddress || !blockchainNetworkId) return

    const response = await handlerQueryErrorCatch(
      AdminQuery(blockchainNetworkId)
    )

    if (response.error) {
      return
    }

    const result = response.result
    setRoyaltyValue(result.marketplaceRoyalty)
    setMarketplaceBalance(result.marketplaceBalance)
  }

  const getSales = async () => {
    if (!userAddress || !blockchainNetworkId) return

    setIsCancelSaleLoading(true)
    const response = await handlerQueryErrorCatch(
      salesQuery(blockchainNetworkId)
    )

    if (response.error) {
      return
    }

    const result = response.result
    setSales(result.sales)
    setIsCancelSaleLoading(false)
    if (result.sales.length > 0) {
      setSaleToCancel(result.sales[0])
    }
  }

  useEffect(() => {
    getRoyaltyValue()
    getSales()
  }, [])

  const updateRoyaltyValue = async () => {
    if (!userAddress || !blockchainNetworkId) return

    setIsNewRoyaltyValueLoading(true)
    setNewRoyalty(blockchainNetworkId, newRoyaltyValue * 100)
      .then(waitTransaction)
      .then(() => {
        setIsNewRoyaltyValueLoading(false)
      })
      .then(getRoyaltyValue)
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      .catch((err: any) => {
        // First phrase of the error message is enough
        const errorMessage = err.message.split('.')[0]
        toast({
          variant: 'solid',
          title: 'Transaction failed',
          description: errorMessage,
          status: 'warning',
          duration: 9000,
          isClosable: true,
        })
        setIsNewRoyaltyValueLoading(false)
      })
  }

  const cleanStorage = async () => {
    if (!userAddress || !blockchainNetworkId) return

    setIsCleanStorageLoading(true)
    cleanMarketplaceStorage(blockchainNetworkId)
      .then(waitTransaction)
      .then(() => {
        setIsCleanStorageLoading(false)
      })
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      .catch((err: any) => {
        // First phrase of the error message is enough
        const errorMessage = err.message.split('.')[0]

        toast({
          variant: 'solid',
          title: 'Transaction failed',
          description: errorMessage,
          status: 'warning',
          duration: 9000,
          isClosable: true,
        })
        setIsCleanStorageLoading(false)
      })
  }

  const onCancelSale = async () => {
    if (!userAddress || !blockchainNetworkId) return
    if (saleToCancel === null) return
    setIsCancelSaleLoading(true)
    cancelSale(
      blockchainNetworkId,
      saleToCancel.contractAddress,
      saleToCancel.tokenId
    )
      .then(waitTransaction)
      .then(() => {
        setIsCancelSaleLoading(false)
      })
      .then(getSales)
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      .catch((err: any) => {
        // First phrase of the error message is enough
        const errorMessage = err.message.split('.')[0]

        toast({
          variant: 'solid',
          title: 'Transaction failed',
          description: errorMessage,
          status: 'warning',
          duration: 9000,
          isClosable: true,
        })
        setIsCancelSaleLoading(false)
      })
  }

  const withdraw = async () => {
    if (!userAddress || !blockchainNetworkId) return

    setIsWithdrawLoading(true)
    withdrawOwner(blockchainNetworkId)
      .then(waitTransaction)
      .then(() => {
        getRoyaltyValue()
        setIsWithdrawLoading(false)
      })
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      .catch((err: any) => {
        // First phrase of the error message is enough
        const errorMessage = err.message.split('.')[0]

        toast({
          variant: 'solid',
          title: 'Transaction failed',
          description: errorMessage,
          status: 'warning',
          duration: 9000,
          isClosable: true,
        })
        setIsWithdrawLoading(false)
      })
  }

  const withdrawToAll = async (address: string) => {
    if (!userAddress || !blockchainNetworkId) return

    setIsWithdrawToLoading(true)
    withdrawTo(blockchainNetworkId, address)
      .then(waitTransaction)
      .then(() => {
        getRoyaltyValue()
        setIsWithdrawToLoading(false)
      })
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      .catch((err: any) => {
        // First phrase of the error message is enough
        const errorMessage = err.message.split('.')[0]

        toast({
          variant: 'solid',
          title: 'Transaction failed',
          description: errorMessage,
          status: 'warning',
          duration: 9000,
          isClosable: true,
        })
        setIsWithdrawToLoading(false)
      })
  }

  const withdrawToAmount = async (address: string, amount: number) => {
    if (!userAddress || !blockchainNetworkId) return

    setIsWithdrawToAmountLoading(true)
    withdrawToAmountService(blockchainNetworkId, address, amount)
      .then(waitTransaction)
      .then(() => {
        getRoyaltyValue()
        setIsWithdrawToAmountLoading(false)
      })
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      .catch((err: any) => {
        // First phrase of the error message is enough
        const errorMessage = err.message.split('.')[0]

        toast({
          variant: 'solid',
          title: 'Transaction failed',
          description: errorMessage,
          status: 'warning',
          duration: 9000,
          isClosable: true,
        })
        setIsWithdrawToAmountLoading(false)
      })
  }

  return {
    royaltyValue,
    marketplaceBalance,
    newRoyaltyValue,
    setNewRoyaltyValue,
    isNewRoyaltyValueLoading,
    updateRoyaltyValue,
    cleanStorage,
    isCleanStorageLoading,
    sales,
    saleToCancel,
    isCancelSaleLoading,
    onCancelSale,
    withdraw,
    isWithdrawLoading,
    withdrawToAll,
    isWithdrawToLoading,
    withdrawToAmount,
    isWithdrawToAmountLoading,
    setSaleToCancel,
  }
}
