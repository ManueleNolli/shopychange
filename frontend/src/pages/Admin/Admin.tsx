import React from 'react'
import {
  Box,
  Center,
  Divider,
  Heading,
  Text,
  Tag,
  Flex,
  VStack,
  NumberInput,
  NumberInputField,
  Select,
  FormControl,
  FormErrorMessage,
  Input,
} from '@chakra-ui/react'
import { colors } from '../../styles/Colors'
import ShadowButton from '../../components/ShadowButton/ShadowButton'
import LoadingButton from '../../components/LoadingButton/LoadingButton'
import { NFT } from '../../types/components/NFT'
import { simplifiedAddress } from '../../utils/SimplifiedAddress/SimplifiedAddress'
import { Field, Form, Formik } from 'formik'
import { validateAddress } from '../../utils/ValidateAddress/ValidateAddress'

import useAdmin from './useAdmin'

export default function Admin() {
  const {
    royaltyValue,
    marketplaceBalance,
    setNewRoyaltyValue,
    isNewRoyaltyValueLoading,
    updateRoyaltyValue,
    cleanStorage,
    isCleanStorageLoading,
    sales,
    isCancelSaleLoading,
    onCancelSale,
    withdraw,
    isWithdrawLoading,
    withdrawToAll,
    isWithdrawToLoading,
    withdrawToAmount,
    isWithdrawToAmountLoading,
    setSaleToCancel,
  } = useAdmin()

  return (
    <Box>
      <Center>
        <Heading>Admin Page</Heading>
      </Center>
      <Box px="2vw" py="2vh">
        <VStack spacing="2vh" align={'start'}>
          <Heading size="md">Royalty</Heading>
          <Flex alignItems={'center'}>
            <Text mr="2vh">Royalty Value</Text>
            {royaltyValue !== null ? (
              <Tag
                size={'lg'}
                variant="solid"
                colorScheme={`${colors.primaryChakraColor}`}
              >
                {royaltyValue} %
              </Tag>
            ) : null}
          </Flex>

          <Flex alignItems={'center'}>
            <Text mr="2vh">Marketplace balance</Text>
            {marketplaceBalance !== null ? (
              <Tag
                size={'lg'}
                variant="solid"
                colorScheme={`${colors.primaryChakraColor}`}
              >
                {marketplaceBalance} ETH
              </Tag>
            ) : null}
          </Flex>

          <Flex alignItems={'center'}>
            <Text mr="2vh">New Royalty value (%)</Text>
            <NumberInput
              mr="2vh"
              w="8em"
              defaultValue={0}
              min={0}
              step={0.01}
              max={100}
              onChange={(value) => setNewRoyaltyValue(parseFloat(value))}
            >
              <NumberInputField
                textAlign="center"
                data-testid="input-royalty"
              />
            </NumberInput>
            <LoadingButton
              isLoading={isNewRoyaltyValueLoading}
              loadingButtonProps={{
                w: '12em',
              }}
            >
              <ShadowButton
                label={'Change Royalty Value'}
                bgColor={`${colors.sellChakraColor}.400`}
                textColor={'white'}
                shadowColor={colors.sellShadowColor}
                hoverColor={`${colors.sellChakraColor}.500`}
                focusColor={`${colors.sellChakraColor}.500`}
                onClick={updateRoyaltyValue}
              />
            </LoadingButton>
          </Flex>

          <Flex alignItems={'center'}>
            <Text mr="2vh">Withdraw</Text>

            <LoadingButton
              isLoading={isWithdrawLoading}
              loadingButtonProps={{
                w: '8em',
              }}
            >
              <ShadowButton
                label={'Withdraw'}
                bgColor={`${colors.sellChakraColor}.400`}
                textColor={'white'}
                shadowColor={colors.sellShadowColor}
                hoverColor={`${colors.sellChakraColor}.500`}
                focusColor={`${colors.sellChakraColor}.500`}
                onClick={withdraw}
              />
            </LoadingButton>
          </Flex>

          <Flex alignItems={'center'}>
            <Text mr="2vh">Withdraw To</Text>

            <Formik
              initialValues={{ address: '' }}
              onSubmit={(values: { address: string }) => {
                withdrawToAll(values.address)
              }}
            >
              {/* eslint-disable-next-line */}
              {({ values }: any) => (
                <Form>
                  <Flex direction={{ base: 'column', md: 'row' }}>
                    <Box w={{ base: '60vw', md: '25vw' }} maxW={'30em'}>
                      <Field
                        name="address"
                        validate={(value: string) => validateAddress(value)}
                      >
                        {/* eslint-disable-next-line */}
                        {({ field, form }: any) => (
                          <FormControl
                            isInvalid={
                              form.errors.address && form.touched.address
                            }
                          >
                            <Input {...field} id={'address'} placeholder="to" />
                            <FormErrorMessage>
                              {form.errors.address}
                            </FormErrorMessage>
                          </FormControl>
                        )}
                      </Field>
                    </Box>
                    <LoadingButton
                      isLoading={isWithdrawToLoading}
                      loadingButtonProps={{
                        w: '12em',
                        ml: { base: '0', md: '2vw' },
                      }}
                    >
                      <ShadowButton
                        label={'Withdraw To'}
                        bgColor={`${colors.sellChakraColor}.400`}
                        textColor={'white'}
                        shadowColor={colors.sellShadowColor}
                        hoverColor={`${colors.sellChakraColor}.500`}
                        focusColor={`${colors.sellChakraColor}.500`}
                        type="submit"
                        buttonProps={{
                          w: '12em',
                          ml: { base: '0', md: '2vw' },
                        }}
                      />
                    </LoadingButton>
                  </Flex>
                </Form>
              )}
            </Formik>
          </Flex>

          <Flex alignItems={'center'}>
            <Text mr="2vh">Withdraw To Amount</Text>

            <Formik
              initialValues={{ address: '', amount: 0 }}
              onSubmit={(values: { address: string; amount: number }) => {
                withdrawToAmount(values.address, values.amount)
              }}
            >
              {/* eslint-disable-next-line */}
              {({ values }: any) => (
                <Form>
                  <Flex direction={{ base: 'column', md: 'row' }}>
                    <Box w={{ base: '60vw', md: '25vw' }} maxW={'30em'}>
                      <Field
                        name="address"
                        validate={(value: string) => validateAddress(value)}
                      >
                        {/* eslint-disable-next-line */}
                        {({ field, form }: any) => (
                          <FormControl
                            isInvalid={
                              form.errors.address && form.touched.address
                            }
                          >
                            <Input {...field} id={'address'} placeholder="to" />
                            <FormErrorMessage>
                              {form.errors.address}
                            </FormErrorMessage>
                          </FormControl>
                        )}
                      </Field>
                    </Box>
                    <Box
                      w={{ base: '20vw', md: '10vw' }}
                      maxW={'10em'}
                      ml={{ base: '0', md: '2vw' }}
                    >
                      <Field key={'amount'} name={'amount'}>
                        {/* eslint-disable-next-line */}
                        {({ field, form }: any) => (
                          <FormControl>
                            <NumberInput
                              min={0}
                              max={marketplaceBalance ? marketplaceBalance : 0}
                              precision={2}
                              step={0.01}
                            >
                              <NumberInputField
                                {...field}
                                id={'amount'}
                                placeholder="Amount"
                              />
                            </NumberInput>
                          </FormControl>
                        )}
                      </Field>
                    </Box>
                    <LoadingButton
                      isLoading={isWithdrawToAmountLoading}
                      loadingButtonProps={{
                        w: '12em',
                        ml: { base: '0', md: '2vw' },
                      }}
                    >
                      <ShadowButton
                        label={'Withdraw To Amount'}
                        bgColor={`${colors.sellChakraColor}.400`}
                        textColor={'white'}
                        shadowColor={colors.sellShadowColor}
                        hoverColor={`${colors.sellChakraColor}.500`}
                        focusColor={`${colors.sellChakraColor}.500`}
                        type="submit"
                        buttonProps={{
                          w: '12em',
                          ml: { base: '0', md: '2vw' },
                        }}
                      />
                    </LoadingButton>
                  </Flex>
                </Form>
              )}
            </Formik>
          </Flex>

          <Divider />
        </VStack>

        <VStack spacing="2vh" align={'start'} mt="2vh">
          <Heading size="md">Maintenance</Heading>
          <Flex alignItems={'center'}>
            <Text mr="2vh">
              Clean Marketplace smartcontract storage (only token{' '}
              <Text as="b">not</Text> for sale)
            </Text>
            <LoadingButton
              isLoading={isCleanStorageLoading}
              loadingButtonProps={{
                w: '8em',
              }}
            >
              <ShadowButton
                label={'Clean Storage'}
                bgColor={`${colors.sellChakraColor}.400`}
                textColor={'white'}
                shadowColor={colors.sellShadowColor}
                hoverColor={`${colors.sellChakraColor}.500`}
                focusColor={`${colors.sellChakraColor}.500`}
                onClick={cleanStorage}
              />
            </LoadingButton>
          </Flex>

          <Flex alignItems={'center'}>
            <Text mr="2vh">Cancel sale</Text>

            <Select
              data-testid="select-sale"
              size="md"
              w={{ base: '50vw', md: '15vw' }}
              mr="2vh"
              onChange={(e) => {
                const selectedSale = sales.find(
                  (sale: NFT) => sale.contractAddress === e.target.value
                )
                if (selectedSale) {
                  setSaleToCancel(selectedSale)
                }
              }}
            >
              {sales.map((sale: NFT) => (
                <option
                  data-testid="option-sale"
                  key={`${sale.contractAddress}.${sale.tokenId}`}
                  value={sale.contractAddress}
                >
                  {simplifiedAddress(sale.contractAddress)} - {sale.tokenId}
                </option>
              ))}
            </Select>

            <LoadingButton
              isLoading={isCancelSaleLoading}
              loadingButtonProps={{
                w: '7em',
              }}
            >
              <ShadowButton
                label={'Cancel Sale'}
                bgColor={`${colors.sellChakraColor}.400`}
                textColor={'white'}
                shadowColor={colors.sellShadowColor}
                hoverColor={`${colors.sellChakraColor}.500`}
                focusColor={`${colors.sellChakraColor}.500`}
                onClick={onCancelSale}
              />
            </LoadingButton>
          </Flex>
          <Divider />
        </VStack>
      </Box>
    </Box>
  )
}
