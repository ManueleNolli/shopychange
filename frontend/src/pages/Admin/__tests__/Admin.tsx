import React from 'react'
import { render, act, fireEvent } from '@testing-library/react'
import Admin from '../Admin'
import useAdmin from '../useAdmin'

jest.mock('../../../utils/SimplifiedAddress/SimplifiedAddress', () => ({
  simplifiedAddress: jest.fn((address: string) => address),
}))

jest.mock('../useAdmin', () => jest.fn())

describe('Admin', () => {
  test('should render the component', async () => {
    ;(useAdmin as jest.Mock).mockReturnValue({
      royaltyValue: 0,
      marketplaceBalance: 0,
      setNewRoyaltyValue: jest.fn(),
      isNewRoyaltyValueLoading: false,
      updateRoyaltyValue: jest.fn(),
      cleanStorage: jest.fn(),
      isCleanStorageLoading: false,
      sales: [],
      isCancelSaleLoading: false,
      onCancelSale: jest.fn(),
      withdraw: jest.fn(),
      isWithdrawLoading: false,
      withdrawToAll: jest.fn(),
      isWithdrawToLoading: false,
      withdrawToAmount: jest.fn(),
      isWithdrawToAmountLoading: false,
      setSaleToCancel: jest.fn(),
    })

    const tree = render(<Admin />)
    expect(tree.container).toMatchSnapshot()
  })

  test('should call change new royalty value', async () => {
    const setNewRoyaltyMock = jest.fn()

    ;(useAdmin as jest.Mock).mockReturnValue({
      royaltyValue: 0,
      marketplaceBalance: 0,
      setNewRoyaltyValue: setNewRoyaltyMock,
      isNewRoyaltyValueLoading: false,
      updateRoyaltyValue: jest.fn(),
      cleanStorage: jest.fn(),
      isCleanStorageLoading: false,
      sales: [],
      isCancelSaleLoading: false,
      onCancelSale: jest.fn(),
      withdraw: jest.fn(),
      isWithdrawLoading: false,
      withdrawToAll: jest.fn(),
      isWithdrawToLoading: false,
      withdrawToAmount: jest.fn(),
      isWithdrawToAmountLoading: false,
      setSaleToCancel: jest.fn(),
    })

    const tree = render(<Admin />)

    const input = tree.getByTestId('input-royalty')
    expect(input).toBeInTheDocument()

    await act(async () => {
      fireEvent.change(input, { target: { value: 10 } })
    })

    expect(setNewRoyaltyMock).toHaveBeenCalledWith(10)
  })

  test('should call submit, witdrawToAll', async () => {
    const withdrawToAllMock = jest.fn()

    ;(useAdmin as jest.Mock).mockReturnValue({
      royaltyValue: 0,
      marketplaceBalance: 0,
      setNewRoyaltyValue: jest.fn(),
      isNewRoyaltyValueLoading: false,
      updateRoyaltyValue: jest.fn(),
      cleanStorage: jest.fn(),
      isCleanStorageLoading: false,
      sales: [],
      isCancelSaleLoading: false,
      onCancelSale: jest.fn(),
      withdraw: jest.fn(),
      isWithdrawLoading: false,
      withdrawToAll: withdrawToAllMock,
      isWithdrawToLoading: false,
      withdrawToAmount: jest.fn(),
      isWithdrawToAmountLoading: false,
      setSaleToCancel: jest.fn(),
    })

    const tree = render(<Admin />)

    const addressInput = tree.queryAllByPlaceholderText('to')[0]

    const submitButton = tree.queryAllByText('Withdraw To')[1]

    await act(async () => {
      fireEvent.change(addressInput, {
        target: { value: '0x965A25dA356e75a58bEBEEA63C3050540259adB2' },
      })
    })

    await act(async () => {
      fireEvent.click(submitButton)
    })

    expect(withdrawToAllMock).toHaveBeenCalled()
  })

  test('should call submit, witdrawToAmount', async () => {
    const withdrawToAmountMock = jest.fn()

    ;(useAdmin as jest.Mock).mockReturnValue({
      royaltyValue: 0,
      marketplaceBalance: 0,
      setNewRoyaltyValue: jest.fn(),
      isNewRoyaltyValueLoading: false,
      updateRoyaltyValue: jest.fn(),
      cleanStorage: jest.fn(),
      isCleanStorageLoading: false,
      sales: [],
      isCancelSaleLoading: false,
      onCancelSale: jest.fn(),
      withdraw: jest.fn(),
      isWithdrawLoading: false,
      withdrawToAll: jest.fn(),
      isWithdrawToLoading: false,
      withdrawToAmount: withdrawToAmountMock,
      isWithdrawToAmountLoading: false,
      setSaleToCancel: jest.fn(),
    })

    const tree = render(<Admin />)

    const addressInput = tree.queryAllByPlaceholderText('to')[1]
    const numberInput = tree.queryAllByPlaceholderText('Amount')[0]

    const submitButton = tree.queryAllByText('Withdraw To Amount')[1]

    await act(async () => {
      fireEvent.change(addressInput, {
        target: { value: '0x965A25dA356e75a58bEBEEA63C3050540259adB2' },
      })

      fireEvent.change(numberInput, {
        target: { value: 10 },
      })
    })

    await act(async () => {
      fireEvent.click(submitButton)
    })

    expect(withdrawToAmountMock).toHaveBeenCalled()
  })

  test('should call submit, witdrawToAmount', async () => {
    const SetSaleToCancelMock = jest.fn()

    ;(useAdmin as jest.Mock).mockReturnValue({
      royaltyValue: 0,
      marketplaceBalance: 0,
      setNewRoyaltyValue: jest.fn(),
      isNewRoyaltyValueLoading: false,
      updateRoyaltyValue: jest.fn(),
      cleanStorage: jest.fn(),
      isCleanStorageLoading: false,
      sales: [
        {
          contractAddress: 'mockContractAddress',
          tokenId: 0,
          image: 'mockImage',
          name: 'mockName',
        },
        {
          contractAddress: 'mockContractAddress1',
          tokenId: 1,
          image: 'mockImage',
          name: 'mockName',
        },
      ],
      isCancelSaleLoading: false,
      onCancelSale: jest.fn(),
      withdraw: jest.fn(),
      isWithdrawLoading: false,
      withdrawToAll: jest.fn(),
      isWithdrawToLoading: false,
      withdrawToAmount: jest.fn(),
      isWithdrawToAmountLoading: false,
      setSaleToCancel: SetSaleToCancelMock,
    })

    const tree = render(<Admin />)

    const select = tree.getByTestId('select-sale')
    expect(select).toBeInTheDocument()

    await act(async () => {
      fireEvent.change(select, {
        target: { value: 'mockContractAddress' },
      })
    })

    expect(SetSaleToCancelMock).toHaveBeenCalledWith({
      contractAddress: 'mockContractAddress',
      tokenId: 0,
      image: 'mockImage',
      name: 'mockName',
    })
  })
})
