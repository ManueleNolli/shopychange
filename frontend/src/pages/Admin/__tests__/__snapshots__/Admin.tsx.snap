// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin should render the component 1`] = `
<div>
  <div
    class="css-0"
  >
    <div
      class="css-gmuwbf"
    >
      <h2
        class="chakra-heading css-0"
      >
        Admin Page
      </h2>
    </div>
    <div
      class="css-serrfu"
    >
      <div
        class="chakra-stack css-16f7yx3"
      >
        <h2
          class="chakra-heading css-0"
        >
          Royalty
        </h2>
        <div
          class="css-70qvj9"
        >
          <p
            class="chakra-text css-uzgswy"
          >
            Royalty Value
          </p>
          <span
            class="css-1qm7lvz"
          >
            0
             %
          </span>
        </div>
        <div
          class="css-70qvj9"
        >
          <p
            class="chakra-text css-uzgswy"
          >
            Marketplace balance
          </p>
          <span
            class="css-1qm7lvz"
          >
            0
             ETH
          </span>
        </div>
        <div
          class="css-70qvj9"
        >
          <p
            class="chakra-text css-uzgswy"
          >
            New Royalty value (%)
          </p>
          <div
            class="chakra-numberinput css-13t5rsu"
          >
            <input
              aria-valuemax="100"
              aria-valuemin="0"
              aria-valuenow="0"
              aria-valuetext="0.00"
              autocomplete="off"
              autocorrect="off"
              class="chakra-numberinput__field css-fzwid6"
              data-testid="input-royalty"
              inputmode="decimal"
              pattern="[0-9]*(.[0-9]+)?"
              role="spinbutton"
              type="text"
              value="0.00"
            />
          </div>
          <button
            class="chakra-button css-79ap6m"
            data-testid="onClick-button"
            style="white-space: initial;"
            type="button"
          >
            Change Royalty Value
          </button>
        </div>
        <div
          class="css-70qvj9"
        >
          <p
            class="chakra-text css-uzgswy"
          >
            Withdraw
          </p>
          <button
            class="chakra-button css-79ap6m"
            data-testid="onClick-button"
            style="white-space: initial;"
            type="button"
          >
            Withdraw
          </button>
        </div>
        <div
          class="css-70qvj9"
        >
          <p
            class="chakra-text css-uzgswy"
          >
            Withdraw To
          </p>
          <form
            action="#"
          >
            <div
              class="css-196weuw"
            >
              <div
                class="css-97mn09"
              >
                <div
                  class="chakra-form-control css-0"
                  role="group"
                >
                  <input
                    class="chakra-input css-0"
                    id="address"
                    name="address"
                    placeholder="to"
                    value=""
                  />
                </div>
              </div>
              <button
                class="chakra-button css-7x4044"
                data-testid="onClick-button"
                style="white-space: initial;"
                type="submit"
              >
                Withdraw To
              </button>
            </div>
          </form>
        </div>
        <div
          class="css-70qvj9"
        >
          <p
            class="chakra-text css-uzgswy"
          >
            Withdraw To Amount
          </p>
          <form
            action="#"
          >
            <div
              class="css-196weuw"
            >
              <div
                class="css-97mn09"
              >
                <div
                  class="chakra-form-control css-0"
                  role="group"
                >
                  <input
                    class="chakra-input css-0"
                    id="address"
                    name="address"
                    placeholder="to"
                    value=""
                  />
                </div>
              </div>
              <div
                class="css-b5ylti"
              >
                <div
                  class="chakra-form-control css-0"
                  role="group"
                >
                  <div
                    class="chakra-numberinput css-e69dqy"
                  >
                    <input
                      aria-readonly="false"
                      aria-required="false"
                      aria-valuemax="0"
                      aria-valuemin="0"
                      autocomplete="off"
                      autocorrect="off"
                      class="chakra-numberinput__field css-8atqhb"
                      id="amount"
                      inputmode="decimal"
                      name="amount"
                      pattern="[0-9]*(.[0-9]+)?"
                      placeholder="Amount"
                      role="spinbutton"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
              <button
                class="chakra-button css-7x4044"
                data-testid="onClick-button"
                style="white-space: initial;"
                type="submit"
              >
                Withdraw To Amount
              </button>
            </div>
          </form>
        </div>
        <hr
          aria-orientation="horizontal"
          class="chakra-divider css-1upb9tn"
        />
      </div>
      <div
        class="chakra-stack css-1wag8h3"
      >
        <h2
          class="chakra-heading css-0"
        >
          Maintenance
        </h2>
        <div
          class="css-70qvj9"
        >
          <p
            class="chakra-text css-uzgswy"
          >
            Clean Marketplace smartcontract storage (only token
             
            <b
              class="chakra-text css-0"
            >
              not
            </b>
             for sale)
          </p>
          <button
            class="chakra-button css-79ap6m"
            data-testid="onClick-button"
            style="white-space: initial;"
            type="button"
          >
            Clean Storage
          </button>
        </div>
        <div
          class="css-70qvj9"
        >
          <p
            class="chakra-text css-uzgswy"
          >
            Cancel sale
          </p>
          <div
            class="chakra-select__wrapper css-he0br4"
          >
            <select
              class="chakra-select css-1ik61og"
              data-testid="select-sale"
            />
            <div
              class="chakra-select__icon-wrapper css-162mkon"
            >
              <svg
                aria-hidden="true"
                class="chakra-select__icon"
                focusable="false"
                role="presentation"
                style="width: 1em; height: 1em; color: currentColor;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </div>
          <button
            class="chakra-button css-79ap6m"
            data-testid="onClick-button"
            style="white-space: initial;"
            type="button"
          >
            Cancel Sale
          </button>
        </div>
        <hr
          aria-orientation="horizontal"
          class="chakra-divider css-1upb9tn"
        />
      </div>
    </div>
  </div>
</div>
`;
